akka-cluster-typed/src/main/scala/akka/cluster/typed/internal/receptionist/Registry.scala

```diff
@@ -0,0 +1,115 @@
+
+  def allDdataKeys: Iterable[DDataKey] = keys
+
+  def ddataKeyFor(serviceKey: ServiceKey[_]): DDataKey =
+    keys(math.abs(serviceKey.hashCode() % serviceRegistries.size))
+
+  def allServices: Iterator[(ServiceKey[_], Set[Entry])] =
+    serviceRegistries.valuesIterator.flatMap(_.map.entries)
+
+  def allEntries: Iterator[Entry] = allServices.flatMap(_._2)
+
+  def getActorRefsFor[T](key: ServiceKey[T]): Set[ActorRef[T]] = {
+    val dDataKey = ddataKeyFor(key)
+    serviceRegistries(dDataKey).getActorRefsFor(key)
+  }
+
+  def withServiceRegistry(dDataKey: DDataKey, registry: ServiceRegistry): SuperServiceRegistry =
+    new SuperServiceRegistry(serviceRegistries + (dDataKey -> registry))
+
+  def allUniqueAddressesInState(selfUniqueAddress: UniqueAddress): Set[UniqueAddress] =
+    allEntries.collect {
+      case entry if entry.ref.path.address.hasGlobalScope ⇒
+        entry.uniqueAddress(selfUniqueAddress)
+    }.toSet
+
+  def collectChangedKeys(dDataKey: DDataKey, newRegistry: ServiceRegistry): Set[AbstractServiceKey] = {
+    val previousRegistry = registryFor(dDataKey)
+    ServiceRegistry.collectChangedKeys(previousRegistry, newRegistry)
+  }
+
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[akka] final case class ServiceRegistry(map: ORMultiMap[ServiceKey[_], Entry]) extends AnyVal {
+
+  // let's hide all the ugly casts we can in here
+  def getActorRefsFor[T](key: AbstractServiceKey): Set[ActorRef[key.Protocol]] =
+    getEntriesFor(key).map(_.ref.asInstanceOf[ActorRef[key.Protocol]])
+
+  def getEntriesFor(key: AbstractServiceKey): Set[Entry] =
+    map.getOrElse(key.asServiceKey, Set.empty[Entry])
+
+  def addBinding[T](key: ServiceKey[T], value: Entry)(implicit cluster: Cluster): ServiceRegistry =
+    ServiceRegistry(map.addBinding(key, value))
+
+  def removeBinding[T](key: ServiceKey[T], value: Entry)(implicit cluster: Cluster): ServiceRegistry =
+    ServiceRegistry(map.removeBinding(key, value))
+
+  def removeAll(removals: Map[AbstractServiceKey, Set[Entry]])(implicit cluster: Cluster): ServiceRegistry = {
+    removals.foldLeft(this) {
+      case (acc, (key, entries)) ⇒
+        entries.foldLeft(acc) {
+          case (innerAcc, entry) ⇒
+            innerAcc.removeBinding[key.Protocol](key.asServiceKey, entry)
+        }
+    }
+  }
+
+  def toORMultiMap: ORMultiMap[ServiceKey[_], Entry] = map
+
+}
+object ServiceRegistry {
+  final val Empty = ServiceRegistry(EmptyORMultiMap)
+
+  def collectChangedKeys(previousRegistry: ServiceRegistry, newRegistry: ServiceRegistry): Set[AbstractServiceKey] = {
+    val allKeys = previousRegistry.toORMultiMap.entries.keySet ++ newRegistry.toORMultiMap.entries.keySet
+    allKeys.foldLeft(Set.empty[AbstractServiceKey]) { (acc, key) ⇒
+      val oldValues = previousRegistry.getEntriesFor(key)
+      val newValues = newRegistry.getEntriesFor(key)
+      if (oldValues != newValues) acc + key
+      else acc
+    }
+  }
+}
```

Comment:
The receptionist just grew and grew so decided to break this out, should add some unit tests for it as well, but didn't get to it yet.

Reply:
:+1: